File: 05/hyeongseok/295.Find_Median_From_Data_Stream.ts
import PriorityQueue

class MedianFinder {
    minHeap;
    maxHeap;
    constructor() {
        this.minHeap = new PriorityQueue({
            compare: (a, b) => a - b,
        })
        this.maxHeap = new PriorityQueue({
            compare: (a, b) => b - a,
        })
    }

    addNum(num: number): void {
        if (this.maxHeap.size() === 0 || this.maxHeap.front() >= num) {
            this.maxHeap.enqueue(num);
        } else {
            this.minHeap.enqueue(num);
        }

        // balancing the two heaps
        if (this.maxHeap.size() > this.minHeap.size() + 1) {
            this.minHeap.enqueue(this.maxHeap.dequeue());
        } else if (this.minHeap.size() > this.maxHeap.size()) {
            this.maxHeap.enqueue(this.minHeap.dequeue());
        }
    }

    findMedian(): number {
        if (this.minHeap.size() == this.maxHeap.size()) {
            return (this.minHeap.front() + this.maxHeap.front()) / 2;
        } else return this.maxHeap.front();
    }
}

File: 05/jaeyoung/295_Find_Median_from_Data_Stream.cpp
class MedianFinder {
    priority_queue<int> max_heap;
    priority_queue<int, vector<int>, greater<int>> min_heap;

   public:
    MedianFinder() {}

    void addNum(int num) {
        if (max_heap.size() == min_heap.size())
            max_heap.push(num);
        else
            min_heap.push(num);

        if (!max_heap.empty() && !min_heap.empty() &&
            max_heap.top() > min_heap.top()) {
            int max_temp = max_heap.top();
            int min_temp = min_heap.top();

            max_heap.pop();
            min_heap.pop();

            max_heap.push(min_temp);
            min_heap.push(max_temp);
        }
    }

    double findMedian() {
        if (max_heap.size() > min_heap.size()) return (double)max_heap.top();

        return (double)(max_heap.top() + min_heap.top()) / 2;
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */